signac/synced_collections/backends/collection_mongodb.py

```diff
@@ -0,0 +1,187 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Implements a MongoDB SyncedCollection backend."""
+from .. import SyncedAttrDict, SyncedCollection, SyncedList
+
+try:
+    import bson
+
+    MONGO = True
+except ImportError:
+    MONGO = False
+
+
+class MongoDBCollection(SyncedCollection):
+    """A :class:`SyncedCollection` that synchronizes with a MongoDB document.
+
+    MongoDB stores documents in collections, where each document is a single
+    record encoded in a JSON-like format (BSON). Each :class:`SyncedCollection` can be
+    represented as a MongoDB document, so this backend stores the :class:`SyncedCollection`
+    as a single document within the collection provided by the user. The document
+    is identified by a unique key provided by the user.
+
+    **Thread safety**
+
+    The MongoDBCollection is not thread-safe.
+
+    Parameters
+    ----------
+    collection : :py:class:`pymongo.collection.Collection`
+        The MongoDB client in which to store data.
+    uid : dict
```

Comment:
We have a special key "data" that cannot exist in the `uid` mapping. This should probably be documented.

Reply:
The default value of `uid` is None in this class and the `MongoDBDict` and `MongoDBList` classes. That seems invalid since this is expecting a `dict`-like object.

Reply:
@vyasr Maybe I'm misunderstanding. I thought that `uid` was meant to be a unique identifier like a hash or primary key integer, except as a dictionary like `{"_id": "abc123def456"}`. From my understanding, `uid=None` will raise an error because it's not a mapping. (Or perhaps it will return an arbitrary record... which would be a bug imo.) That's intentional, then?

Reply:
You're right that `uid` is meant to be a unique identifier. However, a `MongoDBCollection` can contain another `MongoDBCollection` (ad infinitum) and only the top-level collection has a uid, while the uid is None for all others. This is not different than the filename only being set for a top-level `JSONDict`.